File: 2025/generate.rb
#!/usr/bin/env ruby
# Copyright 2025 Trevor Stone
#
# Use of this source code is governed by an MIT-style
# license that can be found in the LICENSE file or at
# https://opensource.org/licenses/MIT.

# Generates a skeletal Advent of Code solution in Ruby.

require 'pathname'
YEAR = 2025
INPUT_BASE = Pathname.new(__FILE__).parent + 'input'
if ARGV.empty?
  STDERR.puts "Usage: #{$PROGRAM_NAME} day1"
  exit 1
end
dayname = ARGV.shift
puts "Generating files in #{dayname}"
daynum = dayname.gsub(/\D+/, '')
dir = Pathname.new dayname
dir.mkdir unless dir.exist?
inputdir = INPUT_BASE + daynum
inputdir.mkdir unless inputdir.exist?

rubyfile = dir + "#{dayname}.rb"
unless rubyfile.exist?
  rubyfile.write <<~ENDCODE
    #!/usr/bin/env ruby
    # Copyright 2025 Trevor Stone
    #
    # Use of this source code is governed by an MIT-style
    # license that can be found in the LICENSE file or at
    # https://opensource.org/licenses/MIT.

    ##
    # Advent of Code #{YEAR} day #{daynum}
    # Read the puzzle at @see https://adventofcode.com/#{YEAR}/day/#{daynum}
    class Day#{daynum}
      def part1 lines
        'TODO'
      end

      def part2 lines
        'TODO'
      end
    end

    if __FILE__ == $PROGRAM_NAME
      require_relative '../runner.rb'
      exit Runner.new.run_day(Day#{daynum}.new, ARGV)
    end
  ENDCODE
  rubyfile.chmod 0755
end

%w(input.actual.txt input.actual.expected).map {|f| dir + f}.each do |f|
  File.symlink Pathname.new('..') + inputdir + f.basename, f unless f.exist? or f.symlink?
end
%w(input.example.txt input.actual.txt).map {|f| dir + f}.each do |f|
  f.write '' unless f.exist?
end
%w(input.example.exmaple input.actual.exmaple).map {|f| dir + f}.each do |f|
  f.write "part1: \npart2: \n" unless f.exist?
end

File: 2025/generate.pl
#!/usr/bin/env -S perl -w
# Copyright 2025 Trevor Stone
#
# Use of this source code is governed by an MIT-style
# license that can be found in the LICENSE file or at
# https://opensource.org/licenses/MIT.

# Generates an Advent of Code solution template in Perl.

use strict;
use feature qw(say);

die "Usage: $0 day1" unless @ARGV;
my $daydir = shift;
my $day;
($day = $daydir) =~ s/\D+//;
my $year = 2025;
if (!-d $daydir) {
  mkdir $daydir or die "Cannot create $daydir: $!";
};
say "Generating files in $daydir";
my $perlfile = "$daydir/$daydir.pl";
if (!-e $perlfile) {
  open my $fh, '>', $perlfile;
  print $fh <<~EOT
  #!/usr/bin/env -S perl -w
  # Copyright 2025 Trevor Stone
  #
  # Use of this source code is governed by an MIT-style
  # license that can be found in the LICENSE file or at
  # https://opensource.org/licenses/MIT.

  # Advent of Code $year day $day
  # Read the puzzle at https://adventofcode.com/$year/day/$day

  use strict;

  sub part1 {
    return 'TODO';
  }

  sub part2 {
    return 'TODO';
  }

  unless (caller) {
    use FindBin qw(\$Bin);
    require "\$Bin/../runner.pl";
  }
  EOT
  ; close($fh);
  chmod 0755, $perlfile;
}
my $inputdir = "input/$day";
mkdir $inputdir or die "Can't make $inputdir: $!" unless -d "input/$day";
foreach my $f (qw(input.actual.txt input.actual.expected)) {
  &touch("$inputdir/$f") unless -e "$inputdir/$f";
  unless (-e "$daydir/$f") {
    my $time = time;
    &touch("$inputdir/$f");
    symlink "../$inputdir/$f", "$daydir/$f";
  }
}
foreach my $f (qw(input.actual.expected input.example.expected)) {
  if (!-e "$daydir/$f" || -z "$daydir/$f") {
    open my $fh, ">", "$daydir/$f";
    print $fh "part1: \npart2: \n";
    close $fh;
  }
}
&touch("$daydir/input.example.txt");

sub touch {
  my $file = shift;
  open TMPFILE, ">", $file and close TMPFILE
    or die "Could not create $file: $!" unless (-e $file);
}
